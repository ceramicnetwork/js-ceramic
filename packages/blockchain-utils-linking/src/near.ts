import { AccountId } from 'caip'
import { AuthProvider } from './auth-provider.js'
import { asOldCaipString, getConsentMessage, LinkProof } from './util.js'
import * as uint8arrays from 'uint8arrays'
import * as nearApiJs from 'near-api-js'
import * as sha256 from '@stablelib/sha256'

const getSignature = async (
  signer: nearApiJs.Signer,
  accountId: string,
  message: any,
  networkId: string
): Promise<string> => {
  const signed = await signer.signMessage(message, accountId, networkId)
  return uint8arrays.toString(signed.signature, 'base64')
}

export class NearAuthProvider implements AuthProvider {
  readonly isAuthProvider = true

  constructor(
    private readonly near: any,
    private readonly accountName: string,
    private readonly chainRef: string
  ) {
    console.warn(
      'WARN: NearAuthProvider is not fully supported. You may encounter issues using this.'
    )
  }

  async authenticate(message: string): Promise<string> {
    const key = await this.near.connection.signer.keyStore.getKey(this.chainRef, this.accountName)
    const signer = await nearApiJs.InMemorySigner.fromKeyPair(this.chainRef, this.accountName, key)
    const digest = sha256.hash(uint8arrays.fromString(message))
    const { signature } = await signer.signMessage(digest, this.accountName, this.chainRef)
    return uint8arrays.toString(signature, 'base16')
  }

  async createLink(did: string): Promise<LinkProof> {
    const key = await this.near.connection.signer.keyStore.getKey(this.chainRef, this.accountName)
    const signer = await nearApiJs.InMemorySigner.fromKeyPair(this.chainRef, this.accountName, key)
    const { message, timestamp } = getConsentMessage(did, true)
    const signature = await getSignature(signer, this.accountName, message, this.chainRef)
    const account = await this.accountId()
    return {
      version: 2,
      type: 'near',
      message,
      signature,
      account: asOldCaipString(account),
      timestamp,
    }
  }

  async accountId(): Promise<AccountId> {
    const key = await this.near.connection.signer.keyStore.getKey(this.chainRef, this.accountName)
    const signer = await nearApiJs.InMemorySigner.fromKeyPair(this.chainRef, this.accountName, key)
    const publicKey = await signer.getPublicKey(this.accountName, this.chainRef)
    const address = uint8arrays.toString(publicKey.data, 'base58btc')
    return new AccountId({
      address: address,
      chainId: `near:${this.chainRef}`,
    })
  }

  withAddress(address: string): AuthProvider {
    return new NearAuthProvider(this.near, address, this.chainRef)
  }
}
