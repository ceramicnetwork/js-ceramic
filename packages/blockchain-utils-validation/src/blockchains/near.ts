import { BlockchainHandler } from '../blockchain-handler.js'
import { LinkProof } from '@ceramicnetwork/blockchain-utils-linking'
import * as uint8arrays from 'uint8arrays'
import crypto from 'crypto'
import nacl from 'tweetnacl'
import { normalizeAccountId } from '@ceramicnetwork/common'

const verifySignature = async (
  pubKey: Uint8Array,
  message: string,
  signature: Uint8Array
): Promise<boolean> => {
  const hash = crypto.createHash('sha256').update(message).digest()
  const hashString = uint8arrays.toString(hash, 'base64')
  const verified = nacl.sign.detached.verify(
    uint8arrays.fromString(hashString, 'base64'),
    signature,
    pubKey
  )
  return verified
}

const namespace = 'near'

export async function validateLink(proof: LinkProof): Promise<LinkProof | null> {
  const pubKey = uint8arrays.fromString(normalizeAccountId(proof.account).address, 'base58btc')
  const msg = proof.message
  const sig = uint8arrays.fromString(proof.signature, 'base64')
  const is_sig_valid = await verifySignature(pubKey, msg, sig)
  return is_sig_valid ? proof : null
}

export const handler: BlockchainHandler = {
  namespace,
  validateLink,
}
