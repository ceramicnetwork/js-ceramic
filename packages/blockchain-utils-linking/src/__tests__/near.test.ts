import { jest } from '@jest/globals'
import { NearAuthProvider } from '../near.js'
import * as nearApiJs from 'near-api-js'

const did = 'did:3:bafysdfwefwe'
const privateKey =
  'ed25519:9hB3onqC56qBSHpHJaE6EyxKPyFxCxzRBkmjuVx6UqXwygvAmFbwnsLuZ2YHsYJqkPTCygVBwXpNzssvWvUySbd'
const chainRef = 'testnet'
const accountName = 'crustykitty.testnet'
const keyPair = nearApiJs.utils.KeyPair.fromString(privateKey)
const keyStore = new nearApiJs.keyStores.InMemoryKeyStore()
keyStore.setKey(chainRef, accountName, keyPair)

const config = {
  keyStore, // instance of InMemoryKeyStore
  networkId: 'testnet',
  nodeUrl: 'fake-address.org',
}

beforeAll(() => {
  global.Date.now = jest.fn().mockImplementation(() => 666000)
})

afterAll(() => {
  jest.clearAllMocks()
})

describe('Blockchain: NEAR', () => {
  describe('createLink', () => {
    test(`create proof for ${chainRef}`, async () => {
      const near = await nearApiJs.connect(config)
      const nearAuthProvider = new NearAuthProvider(near, accountName, chainRef)
      const proof = await nearAuthProvider.createLink(did)
      expect(proof).toMatchSnapshot()
    })
  })

  describe('authenticate', () => {
    test(`create proof for ${chainRef}`, async () => {
      const near = await nearApiJs.connect(config)
      const authProvider = new NearAuthProvider(near, accountName, chainRef)
      const result = await authProvider.authenticate('msg')
      expect(result).toMatchSnapshot()
    })
  })
})
