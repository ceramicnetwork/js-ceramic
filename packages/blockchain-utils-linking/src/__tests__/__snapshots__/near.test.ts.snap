// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Blockchain: NEAR authenticate create proof for testnet 1`] = `"885b8f016c6360641ac9ad47d24e21a7d9db3d6d04febe433bfa67c0c0ef7d67a709254235b7476d45aac11c41ecbcf89e5c407c5fa7d91c37fd597fff436107"`;

exports[`Blockchain: NEAR createLink create proof for testnet 1`] = `
Object {
  "account": "7beiSVU32rh5uqUcGqY3JfkB3y9p6Pc2Q1GFgcrtK65q@near:testnet",
  "message": "Link this account to your identity

did:3:bafysdfwefwe 
Timestamp: 666",
  "signature": "XJFze3Uz3o/AkljTe1vubjsBsFMKq6vQMd8gUegO8W3oN7MbcapNUpjuGhanShsvSSy2tCl6Itp/ESjn9DlvAg",
  "timestamp": 666,
  "type": "near",
  "version": 2,
}
`;
